lib/off_broadway_twitter/producer.ex

```diff
@@ -3,98 +3,148 @@ defmodule OffBroadwayTwitter.Producer do
+        # I got :einval once, but I suppose I was doing something wrong
+        timer = schedule_connection(@connect_in_ms)
 
-  defp get_tweets(tweets_queue, demand, tweets) do
-    {{:value, tweet}, queue} = :queue.out(tweets_queue)
+        {:noreply, [], %{state | conn: conn, connection_timer: timer}}
 
-    get_tweets(queue, demand - 1, [tweet | tweets])
+      {:error, _other} ->
+        {:stop, :stream_stopped_due_unknown_error, state}
+    end
   end
 
   @impl true
-  def handle_info({:data, tweet}, state) do
-    case Jason.decode(tweet) do
-      {:ok, decoded} ->
-        data = Map.fetch!(decoded, "data")
-        meta = Map.delete(data, "text")
-        text = Map.fetch!(data, "text")
-
-        message = %Message{
-          data: text,
-          metadata: meta,
-          acknowledger: {Broadway.NoopAcknowledger, nil, nil}
-        }
-
-        handle_received_messages(%{
-          state
-          | queue: :queue.in(message, state.queue),
-            size: state.size + 1
-        })
+  def handle_info({tag, _socket, _data} = message, state) when tag in [:tcp, :ssl] do
+    send(self(), {:process_stream, state.conn})
 
-      {:error, _} ->
-        IO.puts("error decoding")
-        handle_received_messages(state)
-    end
+    {:noreply, [], %{state | last_message: message}}
   end
 
-  defp connect(opts) do
-    self_pid = self()
-    token = Keyword.fetch!(opts, :twitter_bearer_token)
-    request = Finch.build(:get, @twitter_stream_url_v2, [{"Authorization", "Bearer #{token}"}])
+  @impl true
+  def handle_info({tag, _socket} = message, state) when tag in [:tcp_closed, :ssl_closed] do
+    {:noreply, [], %{state | last_message: message}}
+  end
+
+  defp process_responses(responses, state) do
+    responses
+    |> Enum.reduce(state, fn response, state ->
+      ref = state.request_ref
+
+      case response do
+        {:data, ^ref, tweet} ->
+          case Jason.decode(tweet) do
+            {:ok, decoded} ->
+              data = Map.fetch!(decoded, "data")
+              meta = Map.delete(data, "text")
+              text = Map.fetch!(data, "text")
 
-    spawn_link(fn ->
-      Finch.stream(request, TwitterFinch, nil, fn
-        {:data, data}, _ ->
-          send(self_pid, {:data, data})
+              message = %Message{
+                data: text,
+                metadata: meta,
+                acknowledger: {Broadway.NoopAcknowledger, nil, nil}
+              }
+
+              %{state | tweets: [message | state.tweets]}
+
+            {:error, _} ->
+              IO.puts("error decoding")
+
+              state
+          end
+
+        {:done, ^ref} ->
+          state
 
-        _, _ ->
-          nil
-      end)
+        {_, _ref, _other} ->
+          state
+      end
     end)
+    |> handle_received_messages()
+  end
+
+  @impl true
+  def handle_demand(demand, state) do
+    handle_received_messages(%{state | demand: state.demand + demand})
```

Comment:
Per Slack, yeah, you don't need to track demand because this is a firehose. it needs to be clear in the article though that in 99% of the cases you want to track demand, but in this case it is a firehose, you are just sending it to GenStage buffer